### PR TITLE
[fix](sql_parser) fix convert function

### LINKED ISSUE #19546: [Enhancement] CONVERT(TIMESTAMP "2004-01-22 21:45:33", CHAR(4))
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

doris

mysql> SELECT CONVERT(TIMESTAMP "2004-01-22 21:45:33", CHAR(4));                                  +------------------------------------------+                                                      | CAST('2004-01-22 21:45:33' AS CHARACTER) |                                                      +------------------------------------------+                                                      | 2004-01-22 21:45:33                      |                                                      +------------------------------------------+                                                      1 row in set (0.05 sec)   

mysql

mysql> SELECT CONVERT(TIMESTAMP "2004-01-22 21:45:33", CHAR(4));                                  +---------------------------------------------------+                                             | CONVERT(TIMESTAMP "2004-01-22 21:45:33", CHAR(4)) |                                             +---------------------------------------------------+                                             | 2004                                              |                                             +---------------------------------------------------+                                             1 row in set, 1 warning (0.00 sec)  

### Solution

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by bobhan1**
# Proposed changes

Issue Number: close #19546 

1. add a new rewriterule `RewriteCastToStringRule` to transfer `cast(xx as char(N)/varchar(N))`/`convert(xx, char(N)/varchar(N))` to `substr(cast(xx as char), 1, N)`
2. remove the rewrite procedure from sql parser
 
## Problem summary

Fix the behaviour of `CONVERT()` function. When the target type is varchar or char(N), we transfer `cast(xx as char(N)/varchar(N))` to `substr(cast(xx as char), 1, N)`

before:
```
mysql> SELECT CONVERT(TIMESTAMP "2004-01-22 21:45:33", CHAR(4));
+------------------------------------------+ 
| CAST('2004-01-22 21:45:33' AS CHARACTER) | 
+------------------------------------------+ 
| 2004-01-22 21:45:33                      | 
+------------------------------------------+ 
1 row in set (0.05 sec)
```

after:
```
mysql> SELECT CONVERT(TIMESTAMP "2004-01-02", CHAR(4));
+-----------------------------------------------+
| substr(CAST('2004-01-02' AS CHARACTER), 1, 4) |
+-----------------------------------------------+
| 2004                                          |
+-----------------------------------------------+
1 row in set (0.85 sec)
```
## Checklist(Required)

* [x] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

**Review by Yukang-Lian — COMMENTED** [review]
plz add some regression tests.

**Comment by bobhan1** [comment]
run buildall

**Review by Yukang-Lian — DISMISSED** [review]
LGTM

**Comment by bobhan1** [comment]
run clickbench

**Comment by bobhan1** [comment]
run buildall

File: fe/fe-core/src/main/cup/sql_parser.cup [hunk 1]
```diff
@@ -6402,7 +6402,23 @@ non_pred_expr ::=
   | KW_NULL KW_IS KW_NOT KW_NULL
   {: RESULT = new BoolLiteral(false); :}
   | KW_CONVERT LPAREN expr:e COMMA type_def:targetType RPAREN
-  {: RESULT = new CastExpr(targetType, e); :}
+  {:
+    CastExpr castExpr = new CastExpr(targetType, e);
```
Comment (morningman) [inline fe/fe-core/src/main/cup/sql_parser.cup:6406]: I think this should be implemented as a SQL RewriteRule, better not doing any rewrite in parser.
And we also need to implement it in new query optimizer: Nereids.

**Review by morningman — COMMENTED** [review]
I think this should be implemented as a SQL RewriteRule, better not doing any rewrite in parser.
And we also need to implement it in new query optimizer: Nereids.

**Comment by bobhan1** [comment]
Should this rewrite be added to `FoldConstantsRule`?

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run feut

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.48 seconds
 stream load tsv:          448 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          25.5 seconds inserted 10000000 Rows, about 392K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230529011232_clickbench_pr_152351.html

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.44 seconds
 stream load tsv:          453 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          26.4 seconds inserted 10000000 Rows, about 378K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230529091224_clickbench_pr_152744.html

**Comment by morningman** [comment]
Maybe we also need to implement it in new optimizer?
cc @morrySnow

**Comment by morrySnow** [comment]
please add same rule to Nereids too

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49 seconds
 stream load tsv:          540 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.8 seconds inserted 10000000 Rows, about 335K ops/s
 storage size: 17161948804 Bytes

**Pull request closed without merging**